Support for controlled components via value prop

idolize opened the pull request:
This is a breaking change: "value" is no longer a part of inputAttributes and instead is now a top-level property. To support uncontroleld behavior, a new "defaultValue" prop was added.

Fixes #28 
@moroshko

moroshko commented:
@idolize Thanks a lot for the Pull Request and the updated README!

I have a couple of questions/comments:
* According to React docs for [controlled](https://facebook.github.io/react/docs/forms.html#controlled-components) and [uncontrolled](https://facebook.github.io/react/docs/forms.html#uncontrolled-components) components, `value` and `defaultValue` are `<input>` attributes/props. Why did you decide to take them out of `inputAttributes`?
* If `value` and `defaultValue` live outside of `inputAttributes`, what happens when user specifies both `value` and `inputAttributes.value` (or `defaultValue` and `inputAttributes.defaultValue`)?
* Looks like the [Events playground](http://localhost:3000/examples/dist/index.html#Events playground) example stopped working. My Chrome tab freezes. Does it work for you?
* It would be nice to have tests for the new controlled behaviour.
* As you suggested, we'll have to change the major version. UPGRADE_GUIDE.md would be much appreciated.

Thanks again for your contribution!

idolize commented:
@moroshko Thanks for taking a look, and for the great feedback. To answer your questions:

- I decided to take `value` and `defaultValue` out of `inputProps` because I think of them more as the actual, selected suggestion value instead of just the transient value of the `<input>` field. When using them in conjunction with  `onSuggestionSelected` this is exactly how it works—the `value` of the `<input>` corresponds to the `this.state.value`, which does not have to equal what is held in the `value` prop.
    - This is just my personal opinion, and I do realize it is a breaking change, so if you would prefer keeping the `value`/`defaultValue` props inside `inputAttributes` then that is totally fine by me :+1: 
- Specifying `value` on the `inputAttributes` does nothing, as the actual `value` prop of the `<input>` is set explicitly to `this.state.value` (which is initialized with the top-level `value`/`defaultValue` props).
    - Perhaps using `PropTypes.shape()` instead of `PropTypes.object` for the propType of `inputAttributes` would make this behavior more clear to the developer, but I didn't want to go that far in case the developer wants to pass in some random props to the `<input>` I hadn't thought of (you are using the spread operator to pass through everything)
- Haven't tested this yet, sorry. I am happy to look at it once I get some more time.
- I agree. I'm still pretty new at writing unit tests for React components (I've mostly just used mocha for pure JS testing), so it may take me a bit longer to add these (feel free to write some tests yourself and I'll try to make them pass if they are not currently)
- If you are still okay with moving `value` and `defaultValue` to the top-level props, I will happily add a UPGRADE_GUIDE.md. Just let me know.
- Will do.

idolize commented on src/Autosuggest.js:
```diff
@@ -18,6 +20,10 @@ export default class Autosuggest extends Component { // eslint-disable-line no-s
   }
 
+  componentWillReceiveProps(nextProps) {
+    this.onUpdate(nextProps.value);
```
This should be wrapped in an `if (typeof nextProps.value === 'string')` statement so it doesn't trigger if the developer is using the component in an uncontrolled way and they decide to change some other prop (like `inputAttributes.className`). I will update this in my next commit.

idolize commented:
@moroshko I updated the PR. All feedback should be addressed:

- The Events playground seems to be working fine
- I added some unit tests for the controlled/uncontrolled behavior
- I added a UPGRADE_GUIDE.md
- I got rid of the dist files

Let me know what you think!

dmitry commented:
I believe `UPGRADE_GUIDE` content is better to put into a readme file.

moroshko commented on src/Autosuggest.js:
```diff
@@ -18,6 +20,12 @@ export default class Autosuggest extends Component { // eslint-disable-line no-s
+  componentWillReceiveProps(nextProps) {
+    if (typeof nextProps.value === 'string') {
+      this.onUpdate(nextProps.value);
+    }
```
If we verify the type here, we shouldn't fail silently. I suggest we have something like:
```
else {
  throw new Error('value must be a string. Found:', nextProps.value);
}
```

idolize replied:
@moroshko I added this check not to really ensure typing, but to make sure we don't trigger the `onUpdate` method when `props.value` is `null` or `undefined`—because that can cause broken behavior in the uncontrolled case. So, throwing an error here is not a good idea because it will likewise break people trying to use the uncontrolled behavior.

Perhaps this is less confusing?

```js
if (nextProps.value) {
  // If this component is controlled, then handle the value update
  this.onUpdate(nextProps.value);
}
```

moroshko commented on src/Autosuggest.js:
```diff
@@ -18,6 +20,12 @@ export default class Autosuggest extends Component { // eslint-disable-line no-s
 
+  componentWillReceiveProps(nextProps) {
+    if (typeof nextProps.value === 'string') {
+      this.onUpdate(nextProps.value);
```
What do you think about changing `this.onUpdate()` to `this.updateValue()`? I feel it's a bit more self descriptive this way. (I used `onSomething()` functions in handlers, e.g. `onBlur={this.onInputBlur}`, but here there is no handler.)

idolize replied:
Well, this function is still sort of a handler function, because it happens in response to the `value` changing *or* the uncontrolled input text changing due to user typing. But, it is a purely-internal handler function, and not one that can be passed in via props, so perhaps a different naming scheme does make sense.

What do you think about `handleValueChange` as a name?

moroshko commented:
@idolize Thanks, looks good!

I think I'm fine with leaving `value` and `defaultValue` on the Autosuggest itself. I just didn't want to create  2 potential places where user could have them (Autosuggest itself and `inputAttributes`) to avoid confusion.

I created a `2.0` branch. Could you please change the Pull Request to `2.0` rather than `master`. I'd like to iterate a bit more on this (add more tests, improve docs, etc) before we merge to `master`.

Thanks a lot for your efforts!

moroshko commented:
@idolize Also, all the examples in `/examples` are using uncontrolled components. We need to have at least one example for controlled component. Feel free to create one, or let me know how do you think would be the best way to demonstrate the controlled behaviour.

pathsny commented:
This is pretty awesome. I'm eagerly awaiting this change! One thing I wonder about is what happens when the results are objects rather than strings. Does the initial value get passed to getSuggestionValue to convert the object to a string? or is the "value" prop always a string rather than an object? (I assume it's not).

idolize commented:
Thanks all!

@pathsny I am not using objects as suggestions currently, so I haven't really investigated that deeply, but from reading the docs it appears the use of objects as options always translates to a string-based value anyway?

idolize commented:
@moroshko Is there a way to change the target branch of a PR after opening it? I guess I can close this one and open a new one if that is your preference.

pathsny commented:
@idolize the value object is converted to a string to render using "suggestionValue". I guess the important thing is maintaining the functionality. 

I have an app that needs this, so I can test out the code. But unfortunately your diff only changes the source and not the dist, so I'm confused about how to make it work with my build system which is jspm. But let me try

idolize commented:
@pathsny Thanks for offering to test! In order to make it work I think you may have to check out the code from my fork, run a build with `npm run build`, create a symlink with `npm link`, and then `npm link react-autosuggest` inside your app's directory. At that point your app should be using the code from this PR.

moroshko commented:
@pathsny is right. `getSuggestionValue()` is used only in `focusOnSuggestionUsingKeyboard()`. So the initial suggestion object won't go through `getSuggestionValue()`. We need to fix this, and also add tests for a controlled Autosuggest using object suggestions.

@idolize I'm not sure if it's possible to change the target branch in the PR, so let's just create another one and reference this discussion there.

idolize commented:
Moved to #40
